tailcfg/tailcfg.go

```diff
@@ -574,9 +584,47 @@ func (req *RegisterRequest) Clone() *RegisterRequest {
 		tok := *res.Auth.Oauth2Token
 		res.Auth.Oauth2Token = &tok
 	}
+	if res.DeviceCert != nil {
+		res.DeviceCert = make([]byte, len(req.DeviceCert))
+		copy(res.DeviceCert, req.DeviceCert)
+	}
+	if res.Signature != nil {
+		res.Signature = make([]byte, len(req.Signature))
+		copy(res.Signature, req.Signature)
+	}
 	return res
 }
 
+// WritePlaintextForSigning writes the signed fields (along with the server and machine
+// public keys) to w needed to sign or verify a RegsterRequest.
+func (req *RegisterRequest) WritePlaintextForSigning(
+	w io.Writer, serverURL string, serverPubKey, machinePubKey wgkey.Key) error {
+	tsGob, err := req.Timestamp.GobEncode()
+	if err != nil {
+		return err
+	}
+	if _, err := w.Write(tsGob); err != nil {
+		return err
+	}
+
+	if _, err := w.Write([]byte(serverURL)); err != nil {
```

Comment:
Traditionally io.WriteString but maybe recent compiler optimizations have obviated that.